tools/behavior_simulator/src/main.rs

```diff
@@ -0,0 +1,58 @@
+use std::{io::stdout, path::PathBuf};
+
+use chrono::Local;
+use clap::Parser;
+use color_eyre::{install, Result};
+use fern::{Dispatch, InitError};
+use log::LevelFilter;
+use tokio_util::sync::CancellationToken;
+
+mod cycler;
+mod interfake;
+mod robot;
+mod server;
+mod simulator;
+mod state;
+
+#[derive(Parser)]
+struct Arguments {
+    scenario_file: PathBuf,
+}
+
+fn setup_logger(is_verbose: bool) -> Result<(), InitError> {
```

Comment:
I still vote against logging and for good error messages. IMHO we can treat logging messages similarly as we treat comments: Think twice if you really need them and if you better improve your error handling. The process should be quiet in the happy-path.

Reply:
For programs that simply run to completion, I agree.
But this program runs as a server, errors generated from a specific client connection are not critical errors that would require panicking. Sure, you could collect all the errors and show them collectively at shutdown, but if you don't also show the error right when it happens it can be confusing to use.
I would argue that there is such a thing as a non-happy but also non-critical path.

Also, on quietness: isn't that what log-levels are for?

Regarding the original suggestion: I just copied this piece of code from the hulk binary. I'm open to using another library, but I think we shouldn't use different ones everywhere.

Reply:
If you attach good enough error contexts, it will also be fine to use that rule for long running processes. Depends on the messages if it will be confusing to use.

Log-levels will not help you enabling e.g. debug-level messages only in situations where you wanted to see them. They will spam all the time and most of the time they only fill your disk storage without any benefit.

Reply:
I'm not sure what you mean. If you defer printing error messages until the process stops and there is some error while trying to connect it's super confusing because you get 0 feedback apart from the fact that it just doesn't work.
Killing the server by panicking on error doesn't seem very clean either.

Reply:
Ok, yes, then let me make it more clear: In long running processes, e.g. servers, you will have parts that are shorter lived, e.g. connections. Those shorter lived parts will have a boundary to the long running parts at which you may throw/output the error or collect and aggregate it later. But this is not my point. In this scenario, I still argue that we should not output anything besides the highest-level point at such a boundary where you cannot bubble the error higher up. Instead, this error wrapping should happen with good contexts instead of logging messages. Again, log-levels do not help you here. What you would need are dynamic log-levels that react to whether an error occurred. Or hiding logging messages until an error occurs. This is exactly what error contexts are.

Error contexts are the next generation of logging. IMHO logging should be considered a code smell.

Reply:
I disagree with Hendrik and would like to see logging in the behavior simulator. The central argument i disagree with is that processes should be quiet in the happy-path. For tasks with user interaction like the behavior simulator, the user demands continuous feedback about the state of the program. Imagine a browser which is quiet during downloading files.
When a process runs, it is not possible to distinguish whether the program is quiet because of some kind of failure or whether it is quiet, because it is "happy"

Reply:
You misunderstood what I meant. I mentioned already, that a process should write an error trace when some kind of failure happens. Then the process leaves the happy path. All other feedback to the user should happen via the primary user interface. And there should be a main one to not confuse the user. For the behavior simulator this is the communication interface or Twix. For your browser example this is the browser UI. Logging does not fit here and only brings the mentioned drawbacks to the table. Imagine a browser which is logging every time it does something of a random tab or logs about its HTTP requests while downloading files. As an exercise, start your favorite browser in the command line and observe what they log. I don't see where the behavior simulator or many of our other tools are different?!

Reply:
In a perfect world, sure.
But we don't live in a perfect world. The current workflow with the behavior simulator is to edit the scenario file, restart the simulator/server, then look at twix once that reconnects.
Currently the communication server is only started after the simulation finishes. That will likely be changed in the next iteration because I want to rerun the simulation when the file or parameters change.
But for now there doesn't even exist a channel to the "main interface" during the simulation run.

Imagine the user accidentally broke the exit condition in the script. If there was no output whatsoever, he'd have no idea what is going on? Did the simulation finish yet? Does the communication server hang somehow? Is it an issue with twix reconnecting?
I don't think it's always clear what exactly a "happy" or "unhappy" path is. Soft error states are a thing. From the POV of the simulator, the case above is perfectly fine. Nothing broke. But the user is definitely unhappy with the result, because there is none. Sure, at some point the simulator will run out of memory to store frames in, but that's besides the point.

It doesn't even have to be an infinite loop. It could be as simple as the simulation running a lot slower than expected.
Sure, at some point we could add progress reporting so that the user can see in twix how many frames have been simulated yet and perhaps even start looking at the first few while the rest is still running.

Reply:
> Imagine the user accidentally broke the exit condition in the script.

Failure, not the happy path -> Error context

> Did the simulation finish yet? Does the communication server hang somehow?

If the communication server has an issue, it will not be in the happy path -> Error context

If the behavior simulator has an issue, it will not be in the happy path -> Error context

If there is no output, it computes the simulation. Do not output anything, because everything is working as intended. Happy path. Especially when the behavior simulator runs in 1 second.

Canceling the simulation by the user (maybe because it took so long), left happy path -> Error context with information where it was canceled. I would really like to develop in this use case. Very hard to notice any errors or abnormal things when the process trashes the console output.

> Is it an issue with twix reconnecting?

Issue in Twix, not happy path -> Either error context or UI error message

Reply:
I vote in favor of moving this discussion and decision to next iteration and merge this first iteration asap.